lorien/Misc/Parser.gd

```diff
@@ -0,0 +1,197 @@
+#
+# Grammar
+# 	Used to add a set of other GrammarElement objects which all are tried
+# 	to produce a valid parsing for a given string. The longest parse will be
+# 	returned as the parse of a Grammar object (Top-Down parsing). They
+#	typically form the root-element of any defined Grammar.
+#
+#	Grammar objects are initialized without arguments, and require the user
+#	to add sub-objects to their Gramar.elements array one-by-one. This is
+#	done to facilitate recursive parser elements.
+#
+#
+# GrammarSequence
+#	This class takes a list of other GrammarElement objects and tries to
+#	apply all of them in the defined sequence. A parse is considered
+#	successful only if all elements where mached successfully.
+#
+#	A GrammarSequence object takes three arguments, one optional:
+#	- The name of the symbol this GrammarSequence represents
+#	- The sequence of elements to match as an Array.
+#	- (optional) Boolean indicating a match should be "flattened" after
+#		successful parsing. Flatting means that direct symbols ancestors with
+#		the same name as the GrammarSequence will be inlined, allowing for
+#		easier definition of recursive sequence parsers
+#		(See arglist parser of the StringTemplating class for an example). 
+#
+#
+# GrammarLiteral
+# 	Class matching literal strings. Needs to be an exact match, but whitespace
+#	can be ignored through the flag GrammarLiteral.ignore_whitespace.
+#
+#
+# GrammarRegex
+# 	Same as GrammarLiteral, but uses a regex to perform the matching. If the
+#   regex contains a capture-group, the capture group will be used as the
+#	Symbol-value. Also can ignore whitespace through the 
+#	GrammarRegex.ignore_whitespace option.
+#
+#
+# ParsedSymbol
+#	If a parse was successful, this object will be return the parse-tree.
+#	The most important attributes are:
+#	- name: The symbol name that is assigned GrammarElement that was used for
+#		this node.
+#	- subsymbols: The list of symbols that are contained in this symbol. Only
+#		relevant for a GrammarSequence which will contain a list of other
+#		symbols.
+#	- value: The data that was matched by the Grammar rule.
+
+# -------------------------------------------------------------------------------------------------
+class ParsedSymbol:
+	extends Reference
+
+	var name: String
+	var last_position: int
+	var subsymbols: Array
+	var value
+	
+	func _init(_name: String, _last_position: int, _subsymbols: Array, _value).():
+		name = _name
+		last_position = _last_position
+		subsymbols = _subsymbols
+		value = _value
+	
+	func find_first_subsymbol(symbol_name: String):
+		for st in subsymbols:
+			if st.name == symbol_name:
+				return st
+		return null
+	
+	func _to_string():
+		var subsymbol_strings = PoolStringArray([])
+		for st in subsymbols:
+			subsymbol_strings.append(str(st))
+		return "<T name={name}{value} [{subsymbols}]>".format({
+			"name": name,
+			"value": (" '%s'" % value) if value else "",
+			"subsymbols": subsymbol_strings.join(", ")
+		})
```

Comment:
You said you want the `# -------------------` separators in your code-style. Does that also hold for inner classes like these? And if so, what style should that be?

Reply:
No, that's fine. For small inner classes like these it's easier to read like this. Just add separators between the classes.